Mathlib/Algebra/Vertex/HetVertexOperator.lean

```diff
@@ -0,0 +1,90 @@
+/-
+Copyright (c) 2024 Scott Carnahan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Carnahan
+-/
+import Mathlib.Algebra.BigOperators.Basic
+import Mathlib.Algebra.Module.LinearMap.Basic
+import Mathlib.RingTheory.HahnSeries.Multiplication
+
+/-!
+# Vertex operators
+In this file we introduce heterogeneous vertex operators using Hahn series.  When `R = ℂ`, `V = W`,
+and `Γ = ℤ`, then this is the usual notion of `meromorphic left-moving 2D field`.  The notion we use
+here allows us to consider composites and multiply by multivariable Laurent series.
+## Definitions
+* HetVertexOperator : An `R`-linear map from an `R`-module `V` to `HahnModule Γ W`.
+* The coefficient function as an `R`-linear map.
+## Main results
+* Ext
+## To do:
+* Composition of heterogeneous vertex operators - values are Hahn series on lex order product
+  (needs PR#10781).
+* `HahnSeries Γ R`-module structure on `HetVertexOperator Γ R V W` (needs PR#10846).  This means we
+  can consider products of the form `(X-Y)^n A(X)B(Y)` for all integers `n`, where `(X-Y)^n` is
+  expanded as `X^n(1-Y/X)^n` in `R((X))((Y))`.
+* more API to make ext comparisons easier.
+* formal variable API, e.g., like the `T` function for Laurent polynomials.
+## References
+
+* [R. Borcherds, *Vertex Algebras, Kac-Moody Algebras, and the Monster*][borcherds1986vertex]
+
+-/
+
+noncomputable section
+
+variable {Γ : Type*} [PartialOrder Γ] {R : Type*} {V W : Type*} [CommRing R]
+  [AddCommGroup V] [Module R V] [AddCommGroup W] [Module R W]
+
+/-- A heterogeneous `Γ`-vertex operator over a commutator ring `R` is an `R`-linear map from an
+`R`-module `V` to `Γ`-Hahn series with coefficients in an `R`-module `W`.-/
+abbrev HetVertexOperator (Γ : Type*) [PartialOrder Γ] (R : Type*) [CommRing R]
```

Comment:
Minor point: heterogeneous equality is called `HEq`. What do you think of `HVertexOperator`?

Reply:
That is much better.  I was unsatisfied with `Het`.